gtsam/slam/PlanarProjectionFactor.h

```diff
@@ -0,0 +1,318 @@
+/**
+ * ProjectionFactor, but with pose2 (robot on the floor)
+ *
+ * This factor is useful for high-school robotics competitions,
+ * which run robots on the floor, with use fixed maps and fiducial
+ * markers.
+ *
+ * @see https://www.firstinspires.org/
+ *
+ * @file PlanarProjectionFactor.h
+ * @brief for planar smoothing
+ * @date Dec 2, 2024
+ * @author joel@truher.org
+ */
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/base/Lie.h>
+
+#include <gtsam/geometry/Cal3DS2.h>
+#include <gtsam/geometry/PinholeCamera.h>
+#include <gtsam/geometry/Point3.h>
+#include <gtsam/geometry/Pose2.h>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/geometry/Rot3.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+#include <gtsam/base/numericalDerivative.h>
+
+
+namespace gtsam {
+
+    /**
+     * @class PlanarProjectionFactorBase
+     * @brief Camera projection for robot on the floor.  Measurement is camera pixels.
+     */
+    class PlanarProjectionFactorBase {
+    protected:
+        PlanarProjectionFactorBase() {}
+
+        /**
+         * @param measured corresponding point in the camera frame
+         * @param poseKey index of the robot pose in the z=0 plane
+         */
+        PlanarProjectionFactorBase(const Point2& measured) : measured_(measured) {}
+
+        /**
+         * @param landmark point3
+         * @param pose
+         * @param offset oriented parallel to pose2 zero i.e. +x
+         * @param calib
+         * @param H1 landmark jacobian
+         * @param H2 pose jacobian
+         * @param H3 offset jacobian
+         * @param H4 calib jacobian
+         */
+        Point2 h(
+            const Point3& landmark,
+            const Pose2& pose,
+            const Pose3& offset,
+            const Cal3DS2& calib,
+            OptionalMatrixType H1, // 2x3 (x, y, z)
+            OptionalMatrixType H2, // 2x3 (x, y, theta)
+            OptionalMatrixType H3, // 2x6 (rx, ry, rz, x, y, theta)
+            OptionalMatrixType H4  // 2x9
+        ) const {
+            try {
+                // this is x-forward z-up
+                gtsam::Matrix H0; // 6x6
+                Pose3 offset_pose = Pose3(pose).compose(offset, H0);
+                // this is z-forward y-down
+                gtsam::Matrix H00; // 6x6
+                Pose3 camera_pose = offset_pose.compose(CAM_COORD, H00);
```

Comment:
Oh, I thought we already had wTc. Hmmm, what is CAM_COORD? Where is it defined?

Because the two offsets combine, I actually think the factor would be more general/useful if you just had one bTc that combines offset and extrinsics. Maybe I'm not seeing something.

Style nits:
- use something like kCameraExtrinsics rather than python-style uppercase.
- use camelCase for variables, myVariable.
- But, for poses, use aTc naming convention.

Reply:
The "CAM_COORD" transforms the supplied "bTc" offset so that bTc zero is pointing at +x instead of +z, because that's where FRC code thinks the pose3 zero is, so the caller will want to supply an offset like that.

https://github.com/wpilibsuite/allwpilib/blob/80c391e18278fcb91189a664bb2b2ddaad7eb7a9/wpimath/src/main/java/edu/wpi/first/math/geometry/Pose3d.java#L40

Am I understanding correctly that the camera zero is aiming at +z?  Maybe there's another way handle this transform?

Reply:
Oops, I actually never saw this question. I think what you're saying in code now is:
```c++
camera_pose = FromPose2(wTb) * bTc * CAM_COORD.
```

and you are optimizing over wTb and also, when calibrating, over bTc. Say, in the latter, you start the optimization with bTc_0 and it optimizes to bTc_1. You then use bTc_1 during operation.

OK, well, you can also do this:
```c++
camera_pose = FromPose2(wTb) * new_bTc
```
As a user, you can initialize to new_bTc_0 = bTc_0 * CAM_COORD, and this will then optimize to new_bTc_1 = bTc_1 * CAM_COORD. Then use new_bTc_1 during operation. 

In other words: unless I'm missing something, not something the factor need to be aware of :-)